### PR TITLE
cog 0.5.1 (new formula)

### PR DESCRIPTION
**Pull request by mattt**
Cog is an open-source tool that lets you package machine learning models in a standard, production-ready container. For more information, see https://github.com/replicate/cog.

---

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

### CODE REVIEW

File: Formula/cog.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
+class Cog < Formula
+  desc "Containers for machine learning"
+  homepage "https://github.com/replicate/cog"
+  url "https://github.com/replicate/cog.git",
+      tag:      "v0.5.1",
+      revision: "b58bd63e3cd9b7db0a3410ef47c61c1b11255809"
```
Comment (carlocab) [inline Formula/cog.rb:6]: Do we need a Git clone? Can we use a tarball?

File: Formula/cog.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  license "Apache-2.0"
+
+  head "https://github.com/replicate/cog.git", branch: "main"
+
+  livecheck do
+    url :stable
+    strategy :github_latest
+  end
```
Comment (carlocab) [inline Formula/cog.rb:14]: Does the default livecheck not work?

File: Formula/cog.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+class Cog < Formula
+  desc "Containers for machine learning"
+  homepage "https://github.com/replicate/cog"
+  url "https://github.com/replicate/cog.git",
+      tag:      "v0.5.1",
+      revision: "b58bd63e3cd9b7db0a3410ef47c61c1b11255809"
+  license "Apache-2.0"
+
+  head "https://github.com/replicate/cog.git", branch: "main"
```
Comment (carlocab) [inline Formula/cog.rb:9]:
```suggestion
  license "Apache-2.0"
  head "https://github.com/replicate/cog.git", branch: "main"
```

File: Formula/cog.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    ]
+
+    system "make", *args
+    bin.install "cog"
+
+    generate_completions_from_executable(bin/"cog", "completion")
+  end
+
+  test do
+    assert_match "cog version #{version}", shell_output("#{bin}/cog --version")
```
Comment (carlocab) [inline Formula/cog.rb:33]: Can you propose a more substantial test than checking the version or calling for help? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests. 

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

**Review by carlocab — COMMENTED** [review]
Thanks for the PR. A few comments/questions/suggestions.

File: Formula/cog.rb (hunk 4 above)
Comment (carlocab) [reply Formula/cog.rb:33]: We can try a command with missing dependencies and check that we get the expected error code/message.

Comment (carlocab) [reply Formula/cog.rb:26]: We pass `COG_VERSION` to make, so let's keep the `--version` check (to make sure that the `test` block catches upstream build system changes in specifying the version). But doing `cog build` as well is good.

Unless it exits with `0` you'll need to pass a second argument to `shell_output` specifying the exit code to expect.

File: Formula/cog.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+class Cog < Formula
+  desc "Containers for machine learning"
+  homepage "https://github.com/replicate/cog"
+  url "https://github.com/replicate/cog/archive/refs/tags/v0.5.1.tar.gz"
+  sha256 "53118376c1ddb19d9c4ce866d2b2ebc4bd3899a803268fac0c1e534c23b75843"
+  license "Apache-2.0"
+  head "https://github.com/replicate/cog.git", branch: "main"
+
+  depends_on "go" => :build
+  depends_on "python@3.10" => :build
```
Comment (chenrui333) [inline Formula/cog.rb:10]:
```suggestion
  depends_on "python@3.11" => :build
```

**Review by zeke — COMMENTED** [review]
✅ Cog maintainer here. This looks good to me!

**Review by carlocab — APPROVED** [review]
Thanks!

**Review by chenrui333 — APPROVED** [review]

**Comment by chenrui333** [comment]
@mattt, thanks for your contribution to Homebrew! :tada: 🥇

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you!!!!

**Pull request closed without merging**